Loop unroll release

davors72 opened the pull request.

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+
+  // If the loop unrolling factor is an residual number of iterations we need to
+  // let run the loop for the residual part then let it branch into the unrolled
+  // remaining part. We add add one when calucating the remainder to take into
```
"add add"

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+//
+// 1 - User calls LoopUtils::FullyUnroll or LoopUtils::PartiallyUnroll with a
+// loop they wish to unroll. LoopUtils::CanPerformUnroll is used to
+// validate that a given loop can be unrolled. That method (alongwith the the
```
"alongwith the the"

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  // Save the id of the block before we move it.
+  uint32_t new_merge_id = new_exit_bb->id();
+
+  // Add the block the the list of blocks to add, we want this merge block to be
```
"the the"

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  // right at the start of the new blocks.
+  blocks_to_add_.push_back(std::move(new_exit_bb));
+  ir::BasicBlock* new_exit_bb_raw = blocks_to_add_[0].get();
+  ir::BasicBlock* original_condition_block = loop_condition_block_;
```
Do we need to store this, or can we just store its branch instruction (all it's used for later, so far as I can tell)

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+}
+
+bool LoopUtils::PartiallyUnroll(ir::Loop* loop, size_t factor) {
+  if (factor == 1 || !CanPerformUnroll(loop)) return false;
```
What is `1` here? Is there an enum anywhere you can use to improve readability?

davors72 replied:
The unrolling factor refers to the number of bodies the loop should have, so unrolling by one is not unrolling at all. I'll add a comment

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+          // Some optimization passes have been seen to leave dead phis in the
+          // IR so we check that if a phi is used outside of the loop that the
+          // user
+          // is not dead.
```
formatting

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  // Make sure the continue block is a unconditional branch to the header
+  // block.
+  const ir::Instruction& branch = *loop->GetLatchBlock()->ctail();
+  bool branching_assumtion =
```
assumtion -> assumption ?

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+      ir_context_->GetDominatorAnalysis(&function_, *ir_context_->cfg());
+  opt::DominatorTree& tree = analysis->GetDomTree();
+
+  // Starting the loop header BasicBlock, traverse the dominator tree until we
```
Starting the -> Starting at the

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  opt::DominatorTree& tree = analysis->GetDomTree();
+
+  // Starting the loop header BasicBlock, traverse the dominator tree until we
+  // reach the merge blockand add every node we traverse to the set of blocks
```
blockand -> block and

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+
+// For all instructions in |basic_block| check if the operands used are from a
+// copied
+// instruction and if so swap out the operand for the copy of it.
```
formatting

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+OpBranch %34
+%34 = OpLabel
+%35 = OpSLessThan %11 %25 %10
+OpBranchConditional %35 %36 %27
```
This looks odd to me. Why do we put the `OpBranchConditional` after what looks like one copy of the loop body, but before what loops like a second copy. It seems the `x[i] = 1.0f;` happens twice, once before the conditional and once after. Surely the `OpBranchConditional` shouldn't be inside the unrolled loop, or an I missing something?

davors72 replied:
It's only partially unrolled, the first body has folded its condition and the condition is in the second body. It goes header->condition->store value (body)->iteration count. I think that conditional block should actually be in the first block though.

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+
+  // If partially unrolling the |loop| would leave the loop with too many bodies
+  // for its number of iterations then this method should be used. This method
+  // will duplicate the |loop| compltetely, making the duplicated loop the
```
compltetely -> completely

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  for (ir::Function& f : *c->module()) {
+    LoopUtils loop_utils{&f, c};
+
+    for (ir::Loop& loop : loop_utils.GetLoopDescriptor()) {
```
Do you do any ordering with regards to loop nesting? If you unroll an outer loop won't you produce a large number of new loops which you'll then need to unroll, or does this pass not support nested loops?

davors72 replied:
It doesn't support unrolling inner most loops, if you unroll nested loops you have to go inside out. The CanPerfomUnroll function checks that the loop has no children

Alexander-Johnston replied:
So does this mean that if the loop descriptor iterator gives you the inner most loops first, it will unroll them, but if you get them in outer most first order unrolling won't happen. Am I understanding correctly?

davors72 replied:
Yes, and it does give them in that order

Alexander-Johnston commented on source/opt/loop_unroller.cpp:
```diff
@@ -0,0 +1,915 @@
+  }
+
+  if (changed) return Pass::Status::SuccessWithChange;
+  return Pass::Status::SuccessWithoutChange;
```
`return changed ? Status::SuccessWithChange : Status::SuccessWithoutChange;`

Alexander-Johnston reviewed (changes requested):
This review is mostly about formatting and comments, but includes some implementation questions.

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+%33 = OpLabel
+OpBranch %34
+%34 = OpLabel
+OpBranch %38
```
I know this isn't invalid, but the empty basic blocks are weird

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+%99 = OpLabel
+OpBranch %100
+%100 = OpLabel
+OpBranch %101
```
another double empty basic block

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+%26 = OpIAdd %6 %8 %21
+OpBranch %94
+%94 = OpLabel
+%96 = OpPhi %11 %22 %23 %13 %156
```
Dead Phi

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+%3 = OpVariable %18 Function
+OpBranch %24
+%24 = OpLabel
+%28 = OpPhi %11 %22 %23 %13 %27
```
Dead Phi

Alexander-Johnston reviewed (commented):
Nested loop unrolling seems to be doing some odd things

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+OpStore %28 %27
+%29 = OpIEqual %10 %8 %16
+OpSelectionMerge %30 None
+OpBranchConditional %29 %31 %30
```
This branch is for the `if (i == 5) break`, but isn't breaking out of the loop directly. Instead it's jumping to an empty bb right below, which then branches to the end of the loop. Any idea what's causing this?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+         %11 = OpLabel
+         %26 = OpConvertUToF %19 %38
+         %28 = OpAccessChain %27 %23 %38
+               OpStore %28 %26
```
I only see this store twice in the unrolled version, but looking at the fragment above, for a full unroll the body should be present 4 times, right?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+
+  opt::LoopUnroller loop_unroller;
+  SetDisassembleOptions(SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+//  SinglePassRunAndCheck<opt::LoopUnroller>(text, output, false);
```
uncomment

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+void main() {
+    float out_array[9];
+    for (int i = -10; i < -1; i++) {
+      out_array[i] = i;
```
should be `out_array[i+10] = i;`

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+
+  opt::LoopUnroller loop_unroller;
+  SetDisassembleOptions(SPV_BINARY_TO_TEXT_OPTION_NO_HEADER);
+  // SinglePassRunAndCheck<opt::LoopUnroller>(text, expected, false);
```
dead code?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+                                          &iterations));
+  EXPECT_EQ(iterations, 9);
+  SinglePassRunAndCheck<opt::LoopUnroller>(text, expected, false);
+  //    std::cout << std::get<0>(
```
dead code?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+    for (int i = -10; i < -1; i++) {
+      out_array[i] = i;
+    }
+}
```
this comment is wrong, based on the shader below

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+  //  SinglePassRunAndCheck<opt::LoopUnroller>(text, expected, false);
+  SinglePassRunAndCheck<PartialUnrollerTestPass<2>>(text, expected, false);
+
+  //    std::cout << std::get<0>(
```
dead code?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+    EXPECT_EQ(outer_loop.NumImmediateChildren(), 2);
+  }
+
+  //  std::cout << std::get<0>(
```
dead code?

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+    )";
+
+const std::string output =
+R"()";
```
needs and output

Alexander-Johnston commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -0,0 +1,2206 @@
+  )";
+
+const std::string output =
+R"()";
```
needs an output

Alexander-Johnston reviewed (commented):
Some odd outputs in a few of the tests, including dead Phis. Noticed some incomplete tests too

davors72 committed:
Initial support for loop unrolling.

This patch adds initial support for loop unrolling in the form of a
series of utility classes which perform the unrolling. The pass can
be run with the command spirv-opt --loop-unroll. This will unroll
loops within the module which have the unroll hint set. The unroller
imposes a number of requirements on the loops it can unroll. These are
documented in the comments for the LoopUtils::CanPerformUnroll method in
loop_utils.h. Some of the restrictions will be lifted in future patches.
